### DISCUSSION TITLE
How to get links in rodauth emails with dynamic custom route prefix?

### DISCUSSION THREAD

**Discussion by pboling** [janko/rodauth-rails#296] [Q&A]
The link text I get in my rodauth emails, from all environments, is missing the `host`.

NOTE: The `multi_tenant_path_key` is actually coming through perfectly!

```
If you requested a password reset, please go to
:///my:api/<multi_tenant_path_key>/reset-password?key=la-0-di-1-da-2-di-3-la-4-di-5-da
to reset the password for the account.
```

I do have my mailer setup, for example in development:
```
  host = "localhost"
  port = "3000"
  config.action_mailer.default_url_options = {host:, port:}
```

I'm guessing it is because I have a custom prefix, which is essentially:
```
    prefix { "/my:api/#{multi_tenant_path_key}" }
```

NOTE: I figured out how to route the colon (`:`) in both the Rails route and the Roda route tree, but I don't think that's relevant to this issue, and I only mention it to prevent confusion, as it is very uncommon...

Rails: `namespace "my_api", path: "my:api", defaults: {api: ":api"} do`

Roda: `r.on("my:api") do`

**Comment by janko** [comment]
That's strange, rodauth-rails should still be retrieving the URL host from `config.action_mailer.default_url_options` in the mailer, because it's lacking request context there. Any chance you could reproduce it in a fresh Rails app? Or figure out what's different from the official Rails demo app, where the `:host` is being picked up?

**Reply by pboling** [reply]
Will do, return, and report.

**Reply by pboling** [reply]
@janko I think it is not doing route gen correctly due to the way I override things, which you and I have discussed before.

NOTE: `multi_tenant_path_key` above is actually called `path_key` in my app, and I'll use that from now on for brevity...

You said:

> because it's lacking request context there

The `path_key` only has a way to exist within a request context, except if overridden / set on the `rodauth` instance.  In previous discussions / issues you helped me figure out how to build my Mailer overrides, but I must have something wrong, and I doubt the demo app goes to this extreme...

In my `rodauth_app.rb`:
```
      r.on(:path_key) do |path_key|
        tenant = Tenant.fetch_by_path_key(path_key)
        break unless tenant

        rodauth.path_key = r.env[:path_key] = tenant.path_key
```

And in my `rodauth_main.rb`:
```
    create_verify_account_email do
      RodauthMailer.verify_account(self.class.configuration_name, account_id, request.env[:path_key], verify_account_key_value)
    end
```

In my `rodauth_mailer.rb`:
```
class RodauthMailer < ApplicationMailer
  default to: -> { @rodauth.email_to }, from: -> { @rodauth.email_from }

  def verify_account(name, account_id, path_key, key)
    @rodauth = rodauth(name, account_id, path_key) { @verify_account_key_value = key }
    @account = @rodauth.rails_account

    mail(subject: @rodauth.email_subject_prefix + @rodauth.verify_account_email_subject)
  end
```

I had a PR with some of this multi-tenant work, specs & examples done, and I may resurrect it and try to repro there now that I am much further along.
